internal/server/middleware.go

```diff
@@ -49,63 +50,80 @@ func DatabaseMiddleware(db *gorm.DB) gin.HandlerFunc {
 	}
 }
 
-func DestinationMiddleware() gin.HandlerFunc {
-	return func(c *gin.Context) {
-		uniqueID := c.GetHeader("Infra-Destination")
-		if uniqueID != "" {
-			destinations, err := access.ListDestinations(c, uniqueID, "", &models.Pagination{})
-			if err != nil {
-				return
-			}
+func updateDestinationLastSeenAt(c *gin.Context) error {
+	uniqueID := c.Request.Header.Get("Infra-Destination")
+	if uniqueID == "" {
+		return nil
+	}
 
-			switch len(destinations) {
-			case 0:
-				// destination does not exist yet, noop
-			case 1:
-				destination := destinations[0]
-				// only save if there's significant difference between LastSeenAt and Now
-				if time.Since(destination.LastSeenAt) > time.Second {
-					destination.LastSeenAt = time.Now()
-					if err := access.SaveDestination(c, &destination); err != nil {
-						sendAPIError(c, err)
-						return
-					}
-				}
-			default:
-				sendAPIError(c, fmt.Errorf("multiple destinations found for unique ID %q", uniqueID))
-				return
+	// TODO: use GetDestination(ByUniqueID())
+	destinations, err := access.ListDestinations(c, uniqueID, "", &models.Pagination{})
+	if err != nil {
+		return err
+	}
+
+	switch len(destinations) {
+	case 0:
+		// destination does not exist yet, noop
+		return nil
+	case 1:
+		destination := destinations[0]
+		// only save if there's significant difference between LastSeenAt and Now
+		if time.Since(destination.LastSeenAt) > time.Second {
+			destination.LastSeenAt = time.Now()
+			if err := access.SaveDestination(c, &destination); err != nil {
+				return fmt.Errorf("failed to update destination lastSeenAt: %w", err)
 			}
 		}
+		return nil
+	default:
+		return fmt.Errorf("multiple destinations found for unique ID %q", uniqueID)
+	}
+}
 
-		c.Next()
+// TODO: remove duplicate in access package
+func getDB(c *gin.Context) *gorm.DB {
+	db, ok := c.MustGet("db").(*gorm.DB)
+	if !ok {
+		return nil
 	}
+
+	return db
 }
 
-// AuthenticationMiddleware validates the incoming token
-func AuthenticationMiddleware() gin.HandlerFunc {
+// authenticatedMiddleware is applied to all routes that require authentication.
+// It validates the access key, and updates the lastSeenAt of the user, and
+// possibly also of the destination.
+func authenticatedMiddleware() gin.HandlerFunc {
 	return func(c *gin.Context) {
-		if err := RequireAccessKey(c); err != nil {
+		db := getDB(c)
+		authnUser, err := requireAccessKey(db, c.Request)
+		if err != nil {
 			sendAPIError(c, err)
 			return
 		}
 
+		// TODO: save authnUser in a single key
+		c.Set("key", authnUser.AccessKey)
+		c.Set("identity", authnUser.User)
+
+		if err := updateDestinationLastSeenAt(c); err != nil {
```

Comment:
Bit strange to have this as an unconditional call even when users login, feels like a likely spot to create a bug.

Reply:
Do you mean the name of the function could be misleading?

The behaviour hasn't changed, we previously always looked for the `Infra-Destination` header on all authenticated requests.

I could move the `c.Request.Header.Get("Infra-Destination")` condition to this point, so that the `updateDestinationLastSeen` function is only for the update. Or I could rename the function to something like `handleInfraDestinationHeader`.  What do you think?

Reply:
I like `handleInfraDestinationHeader`, it conveys that it might not be updating anything

Reply:
Sounds good! I've made that change